hail/src/main/scala/is/hail/rvd/RVD.scala

```diff
@@ -114,14 +114,49 @@ class RVD(
   ): RVD = {
     require(newKey.forall(rowType.hasField))
     val nPreservedFields = typ.key.zip(newKey).takeWhile { case (l, r) => l == r }.length
-    require(!isSorted || nPreservedFields > 0 || newKey.isEmpty)
+    val maybeKeys = partitioner.keysIfOneToOne()
+    if (!(!isSorted || nPreservedFields > 0 || newKey.isEmpty || maybeKeys.isDefined)) {
```

Comment:
OK, so, I removed the negation, but the maybeKeys condition is definitely necessary.

I think the central confusion here with both this comment and the other one is that `nPreservedFields` is preserved *in order* and I'm broadening the scan-avoiding operations to those that involve key permutations. It is not true that `nPreservedFields == 0` and `!newKey.isEmpty` imply we must scan, this PR is entirely devoted to handling the subset of situations where we need not scan even though no fields are preserved in order.

Reply:
Okay, I agree it's right now. When I wrote that, `maybeKeys` could be nonempty even if you were completely changing the key, which is why I was thinking a scan was unavoidable. You since added the check that the new key fields are a subset of the old ones, and it all looks right now.